File: agent/src/pcap/worker.rs
/*
 * Copyright (c) 2022 Yunshan Networks
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::{
    fs,
    path::PathBuf,
    sync::{
        atomic::{AtomicBool, AtomicI64, AtomicU64, Ordering},
        Arc, Mutex,
    },
    thread::{self, JoinHandle},
    time::Duration,
};

use dashmap::DashMap;
use log::{debug, error, info, warn};

use super::{
    format_time, get_temp_filename,
    writer::{Writer, WriterCounter},
    Packet, PcapPacket, TapType,
};
use crate::rpc::get_timestamp;
use crate::utils::stats::{Counter, CounterType, CounterValue, RefCountable};
use public::queue::{self, Error};

#[derive(Default)]
pub struct WorkerCounter {
    // statsd:"file_creations
    file_creations: AtomicU64,
    // statsd:"file_closes"
    file_closes: AtomicU64,
    // statsd:"file_rejections"
    file_rejections: AtomicU64,
    // statsd:"file_creation_failures"
    file_creation_failures: AtomicU64,
    // statsd:"file_writing_failures"
    file_writing_failures: AtomicU64,
    // statsd:"written_count"
    written_count: AtomicU64,
    // statsd:"written_bytes"
    written_bytes: AtomicU64,
}

#[derive(Debug, Clone)]
struct WorkerConfig {
    max_concurrent_files: u32,
    max_file_size: u32,
    max_file_period: Duration,
    base_directory: PathBuf,
    writer_buffer_size: u32,
}

pub struct Worker {
    pub index: usize,
    config: WorkerConfig,
    counter: Arc<WorkerCounter>,
    writers: Arc<DashMap<u64, Writer>>,
    packet_receiver: Arc<queue::Receiver<PcapPacket>>,
    thread: Mutex<Option<JoinHandle<()>>>,
    interval: Duration,
    running: Arc<AtomicBool>,
    ntp_diff: Arc<AtomicI64>,
}

impl Worker {
    pub fn new(
        index: usize,
        max_concurrent_files: u32,
        max_file_size: u32,
        max_file_period: Duration,
        base_directory: PathBuf,
        writer_buffer_size: u32,
        packet_receiver: queue::Receiver<PcapPacket>,
        interval: Duration,
        ntp_diff: Arc<AtomicI64>,
    ) -> Self {
        Self {
            index,
            interval,
            config: WorkerConfig {
                max_concurrent_files,
                max_file_size,
                max_file_period,
                base_directory,
                writer_buffer_size,
            },
            counter: Default::default(),
            writers: Arc::new(DashMap::new()),
            packet_receiver: Arc::new(packet_receiver),
            thread: Mutex::new(None),
            running: Arc::new(AtomicBool::new(false)),
            ntp_diff,
        }
    }

    pub fn counter(&self) -> &Arc<WorkerCounter> {
        &self.counter
    }

    pub fn start(&self) {
        if self.running.swap(true, Ordering::SeqCst) {
            debug!("pcap worker has already running");
            return;
        }

        let config = self.config.clone();
        let counter = self.counter.clone();
        let writers = self.writers.clone();
        let packet_receiver = self.packet_receiver.clone();
        let interval = self.interval;

        let ntp_diff = self.ntp_diff.clone();
        // 上层Sender要关闭channel，才调用worker的stop方法
        let thread = thread::spawn(move || loop {
            match packet_receiver.recv(Some(interval)) {
                Ok(PcapPacket::Packet(pkt)) => Self::write_pkt(*pkt, &writers, &config, &counter),
                Err(Error::Timeout) => {
                    let now = get_timestamp(ntp_diff.load(Ordering::Relaxed));
                    Self::clean_timeout_file(now, &writers, &config, &counter);
                }
                Err(Error::Terminated(_, _)) | Ok(PcapPacket::Terminated) => {
                    let now = get_timestamp(ntp_diff.load(Ordering::Relaxed));
                    Self::clean_timeout_file(now, &writers, &config, &counter);
                    break;
                }
            }
        });
        self.thread.lock().unwrap().replace(thread);
    }

    pub fn stop(&self) {
        if !self.running.swap(false, Ordering::SeqCst) {
            debug!("pcap worker has already stopped");
            return;
        }

        if let Some(handle) = self.thread.lock().unwrap().take() {
            let _ = handle.join();
        }
        info!(
            "stop pcap worker ({}) writing to {} files",
            self.index,
            self.writers.len()
        );

        for item in self.writers.iter() {
            let writer = self.writers.remove(item.key()).unwrap().1;
            Self::finish_writer(writer, &self.counter);
        }
    }

    fn clean_timeout_file(
        now: Duration,
        writers: &Arc<DashMap<u64, Writer>>,
        config: &WorkerConfig,
        counter: &WorkerCounter,
    ) {
        for item in writers.iter() {
            if now - item.value().first_pkt_time > config.max_file_period {
                let writer = writers.remove(item.key()).unwrap().1;
                Self::finish_writer(writer, counter);
            }
        }
    }

    fn get_writer_key(dispatcher_id: u32, acl_gid: u16, tap_type: TapType) -> u64 {
        (dispatcher_id as u64) << 32 | (acl_gid as u64) << 16 | u16::from(tap_type) as u64
    }

    fn finish_writer(writer: Writer, worker_counter: &WorkerCounter) {
        let (temp_filename, new_filename) = {
            let Writer {
                temp_filename,
                counter,
                tap_type,
                dispatcher_id,
                first_pkt_time,
                last_pkt_time,
                vtap_id,
                ..
            } = writer;

            let mut new_filename = temp_filename.to_path_buf();
            let first_pkt_time = format_time(first_pkt_time);
            let last_pkt_time = format_time(last_pkt_time);

            new_filename.pop();

            new_filename.push(format!(
                "{}_{:012x}_0_{}_{}.{}.pcap",
                tap_type, dispatcher_id, first_pkt_time, last_pkt_time, vtap_id
            ));

            debug!(
                "finish writing {}, renaming to {}",
                temp_filename.display(),
                new_filename.display()
            );

            worker_counter
                .written_count
                .fetch_add(counter.written_count, Ordering::Relaxed);
            worker_counter
                .written_bytes
                .fetch_add(counter.written_bytes, Ordering::Relaxed);

            (temp_filename, new_filename)
        };
        fs::rename(temp_filename.as_path(), new_filename.as_path()).unwrap_or_else(|err| {
            error!(
                "failed to rename from {} to {} because {:?}",
                temp_filename.display(),
                new_filename.display(),
                err
            );
        });
        worker_counter.file_closes.fetch_add(1, Ordering::Relaxed);
    }

    fn should_close_file(pkt_timestamp: Duration, writer: &Writer, config: &WorkerConfig) -> bool {
        // 距离第一个包时长超过1秒, 且大小超过maxFileSize, 则切换pcap文件
        if pkt_timestamp - writer.first_pkt_time > Duration::from_secs(1)
            && writer.counter.written_bytes as u32 >= config.max_file_size
        {
            return true;
        }
        if pkt_timestamp - writer.first_pkt_time > config.max_file_period {
            return true;
        }

        false
    }

    fn write_pkt(
        meta_pkt: Packet,
        writers: &Arc<DashMap<u64, Writer>>,
        config: &WorkerConfig,
        counter: &WorkerCounter,
    ) {
        let tap_type = meta_pkt.tap_type;
        let acl_gid = meta_pkt.acl_gid;
        let dispatcher_id = meta_pkt.dispatcher_id;
        let vtap_id = meta_pkt.vtap_id;
        let pkt_timestamp = meta_pkt.timestamp();
        let key = Self::get_writer_key(dispatcher_id, acl_gid, tap_type);

        if let Some(writer) = writers.get(&key) {
            if Self::should_close_file(pkt_timestamp, writer.value(), config) {
                let writer = writers.remove(&key).unwrap().1;
                Self::finish_writer(writer, counter);
            }
        }

        if !writers.contains_key(&key) {
            if writers.len() as u32 >= config.max_concurrent_files {
                error!(
                    "max concurrent file ({} files) exceeded",
                    config.max_concurrent_files
                );
                counter.file_rejections.fetch_add(1, Ordering::Relaxed);
                return;
            }

            match Writer::new(
                config.base_directory.as_path(),
                config.writer_buffer_size as usize,
                tap_type,
                dispatcher_id,
                acl_gid,
                vtap_id,
                pkt_timestamp,
            ) {
                Ok(writer) => {
                    counter.file_creations.fetch_add(1, Ordering::Relaxed);
                    writers.insert(key, writer);
                }
                Err(err) => {
                    error!(
                        "failed to create writer for {}: {}",
                        get_temp_filename(
                            config.base_directory.as_path(),
                            acl_gid,
                            tap_type,
                            dispatcher_id,
                            pkt_timestamp,
                            vtap_id
                        )
                        .display(),
                        err
                    );
                    counter
                        .file_creation_failures
                        .fetch_add(1, Ordering::Relaxed);
                    return;
                }
            }
        }

        let mut item = writers.get_mut(&key).unwrap();
        let writer = item.value_mut();
        match writer.write(meta_pkt) {
            Ok(_) => {
                let WriterCounter {
                    written_bytes,
                    written_count,
                } = writer.get_and_reset_stats();

                counter
                    .written_count
                    .fetch_add(written_count, Ordering::Relaxed);
                counter
                    .written_bytes
                    .fetch_add(written_bytes, Ordering::Relaxed);
            }
            Err(err) => {
                warn!(
                    "failed to write packet to {}: {}",
                    writer.temp_filename.display(),
                    err
                );
                counter
                    .file_writing_failures
                    .fetch_add(1, Ordering::Relaxed);
            }
        }
    }
}

impl RefCountable for WorkerCounter {
    fn get_counters(&self) -> Vec<Counter> {
        vec![
            (
                "file_creations",
                CounterType::Counted,
                CounterValue::Unsigned(self.file_creations.load(Ordering::Relaxed)),
            ),
            (
                "file_closes",
                CounterType::Counted,
                CounterValue::Unsigned(self.file_closes.load(Ordering::Relaxed)),
            ),
            (
                "file_rejections",
                CounterType::Counted,
                CounterValue::Unsigned(self.file_rejections.load(Ordering::Relaxed)),
            ),
            (
                "file_creation_failures",
                CounterType::Counted,
                CounterValue::Unsigned(self.file_creation_failures.load(Ordering::Relaxed)),
            ),
            (
                "file_writing_failures",
                CounterType::Counted,
                CounterValue::Unsigned(self.file_writing_failures.load(Ordering::Relaxed)),
            ),
            (
                "written_count",
                CounterType::Counted,
                CounterValue::Unsigned(self.written_count.load(Ordering::Relaxed)),
            ),
            (
                "written_bytes",
                CounterType::Counted,
                CounterValue::Unsigned(self.written_bytes.load(Ordering::Relaxed)),
            ),
        ]
    }
}

File: server/ingester/droplet/pcap/worker.go
/*
 * Copyright (c) 2022 Yunshan Networks
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package pcap

import (
	"fmt"
	"net"
	"os"
	"sync"
	"time"
	"unsafe"

	"github.com/op/go-logging"

	"github.com/deepflowio/deepflow/server/libs/datatype"
	"github.com/deepflowio/deepflow/server/libs/queue"
	"github.com/deepflowio/deepflow/server/libs/zerodoc"
)

const (
	QUEUE_BATCH_SIZE = 1024
	BROADCAST_MAC    = datatype.MacInt(^uint64(0) >> 16)
	BROADCAST_IP     = datatype.IPv4Int(^uint32(0))
)

type WriterKey uint64

func getWriterIpv6Key(ip net.IP, aclGID uint16, tapType zerodoc.TAPTypeEnum) WriterKey {
	ipHash := uint32(0)
	for i := 0; i < len(ip); i += 4 {
		ipHash ^= *(*uint32)(unsafe.Pointer(&ip[i]))
	}
	return WriterKey((uint64(ipHash) << 32) | (uint64(aclGID) << 16) | uint64(tapType))
}

func getWriterKey(tapPort uint32, vtapId, aclGID uint16) WriterKey {
	return WriterKey((uint64(tapPort) << 32) | (uint64(aclGID) << 16) | uint64(aclGID))
}

type WrappedWriter struct {
	*Writer

	tempFilename    string
	firstPacketTime time.Duration
	lastPacketTime  time.Duration

	tapPort uint32
	aclGID  uint16
	vtapId  uint16
	tapType zerodoc.TAPTypeEnum
}

type WorkerCounter struct {
	FileCreations        uint64 `statsd:"file_creations"`
	FileCloses           uint64 `statsd:"file_closes"`
	FileRejections       uint64 `statsd:"file_rejections"`
	FileCreationFailures uint64 `statsd:"file_creation_failures"`
	FileWritingFailures  uint64 `statsd:"file_writing_failures"`
	BufferedCount        uint64 `statsd:"buffered_count"`
	WrittenCount         uint64 `statsd:"written_count"`
	BufferedBytes        uint64 `statsd:"buffered_bytes"`
	WrittenBytes         uint64 `statsd:"written_bytes"`
}

type Worker struct {
	packetQueue queue.QueueReader
	index       int

	maxConcurrentFiles int
	maxFileSize        int64
	maxFilePeriod      time.Duration
	baseDirectory      string

	*WorkerCounter

	writers [datatype.TAP_MAX]map[WriterKey]*WrappedWriter

	writerBufferSize int
	tcpipChecksum    bool

	exiting bool
	exited  bool
	exitWg  *sync.WaitGroup
}

func (m *WorkerManager) newWorker(packetQueueID queue.HashKey) *Worker {
	return &Worker{
		packetQueue: m.packetQueueReaders[packetQueueID],
		index:       int(packetQueueID),

		maxConcurrentFiles: m.maxConcurrentFiles / len(m.packetQueueReaders),
		maxFileSize:        int64(m.maxFileSizeMB) << 20,
		maxFilePeriod:      time.Duration(m.maxFilePeriodSecond) * time.Second,
		baseDirectory:      m.baseDirectory,

		WorkerCounter: &WorkerCounter{},

		writerBufferSize: m.blockSizeKB << 10,
		tcpipChecksum:    m.tcpipChecksum,

		exiting: false,
		exited:  false,
		exitWg:  &sync.WaitGroup{},
	}
}

func tapPortToMacString(tapPort uint32) string {
	return fmt.Sprintf("%012x", tapPort)
}

func tapTypeToString(tapType zerodoc.TAPTypeEnum) string {
	if tapType == 3 {
		return "tor"
	}
	return fmt.Sprintf("isp%d", tapType)
}

func formatDuration(d time.Duration) string {
	return time.Unix(0, int64(d)).Format(TIME_FORMAT)
}

func getTempFilename(tapType zerodoc.TAPTypeEnum, tapPort uint32, firstPacketTime time.Duration, index uint16) string {
	return fmt.Sprintf("%s_%s_0_%s_.%d.pcap.temp", tapTypeToString(tapType), tapPortToMacString(tapPort), formatDuration(firstPacketTime), index)
}

func (w *WrappedWriter) getTempFilename(base string) string {
	return fmt.Sprintf("%s/%d/%s", base, w.aclGID, getTempFilename(w.tapType, w.tapPort, w.firstPacketTime, w.vtapId))
}

func (w *WrappedWriter) getFilename(base string) string {
	return fmt.Sprintf("%s/%d/%s_%s_0_%s_%s.%d.pcap", base, w.aclGID, tapTypeToString(w.tapType), tapPortToMacString(w.tapPort), formatDuration(w.firstPacketTime), formatDuration(w.lastPacketTime), w.vtapId)
}

func (w *Worker) shouldCloseFile(writer *WrappedWriter, packet *datatype.MetaPacket) bool {
	// check for file size and time
	if packet.Timestamp-writer.firstPacketTime > time.Second && writer.FileSize()+int64(writer.BufferSize()) >= w.maxFileSize {
		// 距离第一个包时长超过1秒, 且大小超过maxFileSize, 则切换pcap文件
		return true
	}
	if packet.Timestamp-writer.firstPacketTime > w.maxFilePeriod {
		return true
	}
	return false
}

func (w *Worker) finishWriter(writer *WrappedWriter, newFilename string) {
	writer.Close()
	counter := writer.GetAndResetStats()
	w.BufferedCount += counter.totalBufferedCount
	w.WrittenCount += counter.totalWrittenCount
	w.BufferedBytes += counter.totalBufferedBytes
	w.WrittenBytes += counter.totalWrittenBytes
	log.Debugf("Finish writing %s, renaming to %s", writer.tempFilename, newFilename)
	os.Rename(writer.tempFilename, newFilename)
	w.FileCloses++
}

func (w *Worker) writePacket(packet *datatype.MetaPacket, tapType zerodoc.TAPTypeEnum, aclGID uint16) {
	if w.writers[tapType] == nil {
		w.writers[tapType] = make(map[WriterKey]*WrappedWriter)
	}
	key := getWriterKey(packet.TapPort, packet.VtapId, aclGID)
	writer, exist := w.writers[tapType][key]
	if exist && w.shouldCloseFile(writer, packet) {
		newFilename := writer.getFilename(w.baseDirectory)
		w.finishWriter(writer, newFilename)
		delete(w.writers[tapType], key)
		exist = false
	}
	if !exist {
		writer = w.generateWrappedWriter(tapType, aclGID, packet)
		if writer == nil {
			return
		}
		w.writers[tapType][key] = writer
	}
	if err := writer.Write(packet); err != nil {
		log.Debugf("Failed to write packet to %s: %s", writer.tempFilename, err)
		w.FileWritingFailures++
		return
	}
	counter := writer.GetAndResetStats()
	w.BufferedCount += counter.totalBufferedCount
	w.WrittenCount += counter.totalWrittenCount
	w.BufferedBytes += counter.totalBufferedBytes
	w.WrittenBytes += counter.totalWrittenBytes
	writer.lastPacketTime = packet.Timestamp
}

func (w *Worker) generateWrappedWriter(tapType zerodoc.TAPTypeEnum, aclGID uint16, packet *datatype.MetaPacket) *WrappedWriter {
	if len(w.writers) >= w.maxConcurrentFiles {
		if log.IsEnabledFor(logging.DEBUG) {
			log.Debugf("Max concurrent file (%d files) exceeded", w.maxConcurrentFiles)
		}
		w.FileRejections++
		return nil
	}

	directory := fmt.Sprintf("%s/%d", w.baseDirectory, aclGID)
	if _, err := os.Stat(directory); os.IsNotExist(err) {
		os.MkdirAll(directory, os.ModePerm)
	}
	writer := &WrappedWriter{
		tapType:         tapType,
		aclGID:          aclGID,
		vtapId:          packet.VtapId,
		tapPort:         packet.TapPort,
		firstPacketTime: packet.Timestamp,
		lastPacketTime:  packet.Timestamp,
	}

	writer.tempFilename = writer.getTempFilename(w.baseDirectory)
	if log.IsEnabledFor(logging.DEBUG) {
		log.Debugf("Begin to write packets to %s", writer.tempFilename)
	}
	var err error
	if writer.Writer, err = NewWriter(writer.tempFilename, w.writerBufferSize, w.tcpipChecksum); err != nil {
		if log.IsEnabledFor(logging.DEBUG) {
			log.Debugf("Failed to create writer for %s: %s", writer.tempFilename, err)
		}
		w.FileCreationFailures++
		return nil
	}
	w.FileCreations++
	return writer
}

func (w *Worker) cleanTimeoutFile(timeNow time.Duration) {
	for i := datatype.TAP_MIN; i < datatype.TAP_MAX; i++ {
		for key, writer := range w.writers[i] {
			if timeNow-writer.firstPacketTime > w.maxFilePeriod {
				newFilename := writer.getFilename(w.baseDirectory)
				w.finishWriter(writer, newFilename)
				delete(w.writers[i], key)
			}
		}
	}
}

func (w *Worker) toZerodocTAPType(packet *datatype.MetaPacket) zerodoc.TAPTypeEnum {
	if packet.TapType != datatype.TAP_CLOUD {
		return zerodoc.TAPTypeEnum(packet.TapType)
	}
	return zerodoc.CLOUD
}

func (w *Worker) Process() {
	elements := make([]interface{}, QUEUE_BATCH_SIZE)

WORKING_LOOP:
	for !w.exiting {
		n := w.packetQueue.Gets(elements)
		timeNow := time.Duration(time.Now().UnixNano())
		for _, e := range elements[:n] {
			if e == nil { // tick
				if w.exiting {
					break WORKING_LOOP
				}
				w.cleanTimeoutFile(timeNow)
				continue
			}

			block := e.(*datatype.MetaPacketBlock)

			for i := uint8(0); i < block.Count; i++ {
				packet := &block.Metas[i]

				if !packet.EndpointData.Valid() { // shouldn't happen
					log.Warningf("drop invalid packet with nil EndpointData %v", packet)
					continue
				}

				tapType := w.toZerodocTAPType(packet)
				for _, policy := range packet.PolicyData.NpbActions {
					// NOTICE: PCAP存储必须满足TunnelType是NPB_TUNNEL_TYPE_PCAP, 因为策略是NPB_TUNNEL_TYPE_PCAP类型，这里的判断去掉了
					if policy.TunnelGid() <= 0 {
						continue
					}
					w.writePacket(packet, tapType, policy.TunnelGid())
				}
			}

			datatype.ReleaseMetaPacketBlock(block)
		}
	}

	for i := datatype.TAP_MIN; i < datatype.TAP_MAX; i++ {
		for _, writer := range w.writers[i] {
			newFilename := writer.getFilename(w.baseDirectory)
			w.finishWriter(writer, newFilename)
		}
	}
	log.Infof("Stopped pcap worker (%d)", w.index)
	w.exitWg.Done()
}

func (w *Worker) Close() error {
	log.Infof("Stop pcap worker (%d) writing to %d files", w.index, len(w.writers))
	w.exitWg.Add(1)
	w.exiting = true
	w.exitWg.Wait()
	w.exited = true
	return nil
}

func (w *Worker) GetCounter() interface{} {
	counter := &WorkerCounter{}
	counter, w.WorkerCounter = w.WorkerCounter, counter
	return counter
}

func (w *Worker) Closed() bool {
	return w.exited
}
